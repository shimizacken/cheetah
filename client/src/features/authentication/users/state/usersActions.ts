import {Members} from '../../../../packages/socket/savannah.types';
import {User} from '../../../../state/store.types';
import {
  ADD_USER,
  REMOVE_USER,
  SET_USER_ACTIVE,
  SET_CURRENT_USER_ID,
  POST_USER,
  SIGN_OUT
} from './usersConstants';

export const postUser = (user: User) => ({
  type: POST_USER,
  user
});

export const addUser = (user: User) => ({
  type: ADD_USER,
  user
});

export const updateMembers = (members: Members) => ({
  type: 'UPDATE_MEMBERS',
  members
});

export const removeUser = (userId: string) => ({
  type: REMOVE_USER,
  userId
});

export const setUserActive = (userId: string) => ({
  type: SET_USER_ACTIVE,
  userId
});

export const setCurrentUserId = (currentUserId: string) => ({
  type: SET_CURRENT_USER_ID,
  currentUserId
});

export const signOut = (userId: string) => ({
  type: SIGN_OUT,
  userId
});
