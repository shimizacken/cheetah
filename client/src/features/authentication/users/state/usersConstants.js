export const ADD_USER = 'users/add';
export const REMOVE_USER = 'users/remove';
export const SET_USER_ACTIVE = 'users/set/active';

export const SET_CURRENT_USER_ID = 'users/current/id';
