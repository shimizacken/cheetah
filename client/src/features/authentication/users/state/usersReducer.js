const usersInitialState = [];

export const users = (state = usersInitialState) => {
  return state;
};
