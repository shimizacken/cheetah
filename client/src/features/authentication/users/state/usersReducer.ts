import {AnyAction, combineReducers} from 'redux';
import {
  ADD_USER,
  // SET_USER_ACTIVE,
  SET_CURRENT_USER_ID,
  SIGN_OUT
} from './usersConstants';

const usersInitialState = {};

export const users = (state = usersInitialState, action: AnyAction) => {
  if (action.type === 'UPDATE_MEMBERS') {
    return action.members;
  }

  if (action.type === ADD_USER) {
    return {
      ...state,
      ...action.user
    };
  }

  return state;
};

const currentUserIdInitialState = '';

export const currentUserId = (
  state = currentUserIdInitialState,
  action: AnyAction
) => {
  if (action.type === SET_CURRENT_USER_ID) {
    return action.currentUserId;
  }

  if (action.type === SIGN_OUT) {
    return currentUserIdInitialState;
  }

  return state;
};

export const authentication = combineReducers({
  users,
  currentUserId
});
