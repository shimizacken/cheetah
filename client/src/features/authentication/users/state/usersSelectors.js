import { createSelector } from 'reselect';

export const selectUsers = (state) => state?.authentication?.users;

export const selectCurrentUserId = (state) => state?.authentication?.currentUserId;

export const selectUser = (userId) => createSelector(selectUsers, (users) => users?.[userId]);

export const selectIsUserOnline = createSelector(
  selectUsers,
  selectCurrentUserId,
  (users, currentUserId) => users?.[currentUserId] !== undefined
);

export const selectIsCurrentUser = (userId) =>
  createSelector(selectCurrentUserId, (currentUserId) => {
    return userId === currentUserId;
  });
