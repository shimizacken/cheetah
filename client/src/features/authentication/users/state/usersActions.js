import { ADD_USER, REMOVE_USER, SET_USER_ACTIVE, SET_CURRENT_USER_ID, POST_USER } from './usersConstants';

export const postUser = (user) => ({
  type: POST_USER,
  user
});

export const addUser = (user) => ({
  type: ADD_USER,
  user
});

export const removeUser = (userId) => ({
  type: REMOVE_USER,
  userId
});

export const setUserActive = (userId) => ({
  type: SET_USER_ACTIVE,
  userId
});

export const setCurrentUserId = (currentUserId) => ({
  type: SET_CURRENT_USER_ID,
  currentUserId
});
