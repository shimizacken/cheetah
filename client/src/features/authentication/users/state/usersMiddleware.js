import { getWS } from '../bll/network/usersLoader';
import { POST_USER, SIGN_OUT } from './usersConstants';
import { selectUser } from './usersSelectors';

export const usersMiddleware = ({ getState }) => (next) => (action) => {
  if (action.type === POST_USER) {
    getWS().send(JSON.stringify(action.user));
  }

  if (action.type === SIGN_OUT) {
    const user = selectUser(action.userId)(getState());

    getWS().send(
      JSON.stringify({
        ...user,
        active: false
      })
    );
  }

  return next(action);
};
