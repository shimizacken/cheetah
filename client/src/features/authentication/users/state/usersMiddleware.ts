import {Dispatch} from 'react';
import type {MiddlewareAPI, AnyAction, Middleware} from 'redux';
import {getSocket} from '../../../../packages/socket/src/socket';
import {POST_USER, SIGN_OUT} from './usersConstants';
import {selectUser} from './usersSelectors';

export const usersMiddleware: Middleware = ({getState}: MiddlewareAPI) => (
  next: Dispatch<AnyAction>
) => (action: AnyAction) => {
  if (action.type === POST_USER) {
    const socket = getSocket();

    socket.send(
      JSON.stringify({
        type: 'authentication',
        ...action.user
      })
    );
  }

  if (action.type === SIGN_OUT) {
    const socket = getSocket();
    const user = selectUser(action.userId)(getState());

    socket.send(
      JSON.stringify({
        type: 'authentication',
        ...user,
        active: false
      })
    );
  }

  return next(action);
};
