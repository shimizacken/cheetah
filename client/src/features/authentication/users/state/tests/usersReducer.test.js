import { randomAction } from '../../../../../state/randomAction';
import { users } from '../usersReducer';
import { addUser, setUserActive } from '../usersActions';

describe('usersReducer test suite', () => {
  describe('defaults', () => {
    it('should return default state', () => {
      const action = randomAction();

      expect(users(undefined, action)).toMatchSnapshot();
    });
  });

  describe('add user test suite', () => {
    it('should add first user', () => {
      const state = {};
      const user = {
        id: '1',
        displayName: 'jimi zax',
        active: false
      };

      expect(users(state, addUser(user))).toMatchSnapshot();
    });

    it('should add another user', () => {
      const state = {
        '1': {
          id: '1',
          displayName: 'jimi',
          active: false
        }
      };
      const user = {
        id: '2',
        displayName: 'leah',
        active: false
      };

      expect(users(state, addUser(user))).toMatchSnapshot();
    });
  });

  describe.only('set user active test suite', () => {
    it('should activate user', () => {
      const userId = '1';
      const state = {
        [userId]: {
          id: '1',
          displayName: 'jimi',
          active: false
        },
        '2': {
          id: '2',
          displayName: 'leah',
          active: false
        }
      };

      expect(users(state, setUserActive(userId))).toMatchSnapshot();
    });

    it('should deactivate user', () => {
      const userId = '1';
      const state = {
        [userId]: {
          id: '1',
          displayName: 'jimi',
          active: true
        },
        '2': {
          id: '2',
          displayName: 'leah',
          active: false
        }
      };

      expect(users(state, setUserActive(userId))).toMatchSnapshot();
    });
  });
});
