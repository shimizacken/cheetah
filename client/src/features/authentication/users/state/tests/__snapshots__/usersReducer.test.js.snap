// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`usersReducer test suite add user test suite should add another user 1`] = `
Object {
  "1": Object {
    "active": false,
    "displayName": "jimi",
    "id": "1",
  },
  "active": false,
  "displayName": "leah",
  "id": "2",
}
`;

exports[`usersReducer test suite add user test suite should add first user 1`] = `
Object {
  "active": false,
  "displayName": "jimi zax",
  "id": "1",
}
`;

exports[`usersReducer test suite defaults should return default state 1`] = `Object {}`;

exports[`usersReducer test suite set user active test suite should activate user 1`] = `
Object {
  "1": Object {
    "active": true,
    "displayName": "jimi",
    "id": "1",
  },
  "2": Object {
    "active": false,
    "displayName": "leah",
    "id": "2",
  },
}
`;

exports[`usersReducer test suite set user active test suite should deactivate user 1`] = `
Object {
  "1": Object {
    "active": false,
    "displayName": "jimi",
    "id": "1",
  },
  "2": Object {
    "active": false,
    "displayName": "leah",
    "id": "2",
  },
}
`;
